src/Model/FileTag.php

```diff
@@ -11,127 +11,149 @@
-     * @param string $type  Optional file type.
-     * 
-     * @return string       Comma delimited list of tag names.
-     */
-    public static function fileToList($file, $type = 'file')
-    {
-        $matches = false;
-        $list = '';
-
-        if ($type == 'file') {
-            $cnt = preg_match_all('/\[(.*?)\]/', $file, $matches, PREG_SET_ORDER);
-        } else {
-            $cnt = preg_match_all('/<(.*?)>/', $file, $matches, PREG_SET_ORDER);
-        }
-
-        if ($cnt)
-        {
-            foreach ($matches as $mtch)
-            {
-                if (strlen($list))
-                {
-                    $list .= ',';
-                }
-
-                $list .= self::decode($mtch[1]);
-            }
-        }
-
-        return $list;
-    }
+	/**
+	 * @brief URL encode <, >, left and right brackets
+	 *
+	 * @param string $s String to be URL encoded.
+	 *
+	 * @return string   The URL encoded string.
+	 */
+	public static function encode($s)
+	{
+		return str_replace(['<', '>', '[', ']'], ['%3c', '%3e', '%5b', '%5d'], $s);
+	}
+
+	/**
+	 * @brief URL decode <, >, left and right brackets
+	 *
+	 * @param string $s The URL encoded string to be decoded
+	 *
+	 * @return string   The decoded string.
+	 */
+	public static function decode($s)
+	{
+		return str_replace(['%3c', '%3e', '%5b', '%5d'], ['<', '>', '[', ']'], $s);
+	}
+
+	/**
+	 * @brief Query files for tag
+	 *
+	 * @param string $table The table to be queired.
+	 * @param string $s     The search term
+	 * @param string $type  Optional file type.
+	 *
+	 * @return string       Query string.
+	 */
+	public static function fileQuery($table, $s, $type = 'file')
+	{
+		if ($type == 'file') {
+			$str = preg_quote('[' . str_replace('%', '%%', self::encode($s)) . ']');
+		} else {
+			$str = preg_quote('<' . str_replace('%', '%%', self::encode($s)) . '>');
+		}
+
+		return " AND " . (($table) ? DBA::escape($table) . '.' : '') . "file regexp '" . DBA::escape($str) . "' ";
+	}
+
+	/**
+	 * Get file tags from array
+	 *
+	 * ex. given [music,video] return <music><video> or [music][video]
+	 *
+	 * @param array  $array A list of tags.
+	 * @param string $type  Optional file type.
+	 *
+	 * @return string       A list of file tags.
+	 */
+	public static function arrayToFile(array $array, string $type = 'file')
+	{
+		$tag_list = '';
+		if ($type == 'file') {
+			$lbracket = '[';
+			$rbracket = ']';
+		} else {
+			$lbracket = '<';
+			$rbracket = '>';
+		}
+
+		foreach ($array as $item) {
+			if (strlen($item)) {
+				$tag_list .= $lbracket . self::encode(trim($item)) . $rbracket;
+			}
+		}
+
+		return $tag_list;
+	}
+
+	/**
+	 * Get tag list from file tags
+	 *
+	 * ex. given <music><video>[friends], return [music,video] or [friends]
+	 *
+	 * @param string $file File tags
+	 * @param string $type Optional file type.
+	 *
+	 * @return array        List of tag names.
+	 */
+	public static function fileToArray(string $file, string $type = 'file')
```

Comment:
Maybe you can add some test cases, you probably already tested with?

Reply:
Here you go.